### PR TITLE
DR-2494: Add DRS post endpoints to support RAS passport Auth

### PR DESCRIPTION
**Pull request by snf2ye**
Note to reviewers:
- I broke down the changes into meaningful commits.  Otherwise, the changes associated with a relatively small change (adding a directory "auth" to store "iam" and "ras") overwhelmed the PR. 
- I've added comments for areas that I'd particularly like feedback on
- I'm least confident about how I wired in the ECMService, so I would definitely appreciate feedback there

### Change Log
- Utilize the External Credentials Manager (ECM) to validate passports. This means we're working with a new service. Follow-on PRs will be needed to reference the ECM URL for other environments (alpha, staging, prod, etc.) to our helm definitions repo.
- Add two new endpoints: PostObjectId and PostAccessURL
- Attempt to gracefully share the bulk of the DRSService code between GetObjectId & PostObjectID and then between GetAccessURL and PostAccessURL.
- Added parent "auth" folder to store IAM and ECM code. Definitely would appreciate feedback on whether this makes sense - I wasn't sure where else to put the ECM service code. For the subfolders - Would a "ecm" folder make more sense than a "ras" folder?
- Added unit tests that should sufficiently confirm that we are throwing an exception when the user provides an invalid passport or when the passport does not align with snapshot's PHSID/consent code.
- Add audit logging for response from ecm validate passport call
- **Potentially Breaking Change:** For the PostObject() endpoint, I'm only returning an access_id, NOT an access_url. I've changed the behavior for Azure to match this behavior. I've reached out on slack to users that this might potentially impact and am awaiting feedback. So far, the batch team seems okay with this change. 

**Other add-on changes:**
- We referenced "region.getValue()" and "region.getString()" seemingly interchangeably. As far as I can tell, they return the same value, so I've made it so we always use 

### Background
* More info on PHSIds and consent codes in [this doc](https://docs.google.com/document/d/10OW4Em_x88E8RxGASwc-tnlFbLtqdBZ5ewLowM49xYQ/edit).

The overarching idea of this PR is illustrated in the diagram below: We want the user to be able to provide a RAS passport and a DRS Object Id and get back a signed url.
TDR File Access

Following the DRS spec, this is done in two steps:
DRS PostObject Endpoint

### CODE REVIEW

File: src/main/java/bio/terra/app/configuration/ApplicationConfiguration.java [hunk 1]
```diff
@@ -397,6 +399,11 @@ public ExecutorService performanceThreadpool() {
         new LinkedBlockingQueue<>(getMaxPerformanceThreadQueueSize()));
   }
 
+  @Bean
+  public RestTemplate restTemplate(RestTemplateBuilder builder) {
+    return builder.build();
+  }
+
```
Comment (snf2ye) [inline src/main/java/bio/terra/app/configuration/ApplicationConfiguration.java:406]: I don't really understand why we seemingly need this RestTemplate bean. It appears we reference a rest template without the bean w/ some of our other referenced services.

File: src/main/java/bio/terra/service/auth/ras/ECMService.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    // they should be parsed as JSON.
+    // N.B. This means that any non-JSON responses won't be parsed correctly, so if ECM ever returns
+    // any other content-type, this would have to be revised. However, that seems unlikely.
+    var restTemplate = new RestTemplate(List.of(jacksonConverter));
+
+    var client = new ApiClient(restTemplate);
+    client.setBasePath(ecmConfiguration.getInstanceUrl());
+
+    return new PassportApi(client);
+  }
```
Comment (snf2ye) [inline src/main/java/bio/terra/service/auth/ras/ECMService.java]: This is all stolen from the DRSHub repo. Not sure if there is an easier way for us to reference this service.

File: src/main/java/bio/terra/service/filedata/DrsService.java [hunk 3]
```diff
@@ -146,69 +180,122 @@ public void close() {
...
+    var criteria = new RASv1Dot1VisaCriterion().consentCode(consentCode).phsId(phsId);
+    criteria.issuer("https://stsstg.nih.gov").type("RASv1Dot1VisaCriterion");
+    var request =
+        new ValidatePassportRequest()
+            .passports(drsPassportRequestModel.getPassports())
+            .criteria(List.of(criteria));
+    var result = ecmService.validatePassport(request);
+
+    // log audit info
+    logger.info(result.getAuditInfo().toString());
```
Comment (snf2ye) [inline src/main/java/bio/terra/service/filedata/DrsService.java]: @nmalfroy  - Is this sufficient for the audit logging?

File: src/test/java/bio/terra/service/tabulardata/google/BigQueryPdaoTest.java [hunk 4]
```diff
@@ -25,11 +25,11 @@
 import bio.terra.model.IngestRequestModel;
 import bio.terra.model.SnapshotModel;
 import bio.terra.model.SnapshotSummaryModel;
+import bio.terra.service.auth.iam.IamProviderInterface;
```
Comment (nmalfroy) [inline src/test/java/bio/terra/service/tabulardata/google/BigQueryPdaoTest.java:28]: FWIW, I agree with your naming.  Nesting in an `auth` package makes a ton of sense

File: src/main/java/bio/terra/app/configuration/ApplicationConfiguration.java (hunk 1 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/app/configuration/ApplicationConfiguration.java:406]: How do you mean?  Was it complaining about not having a bean?  Could be that the ECM client requires it?  FWIW, not a big deal since I'm about to put up a PR that needs a Rest template bean :)

File: src/main/java/bio/terra/app/configuration/ECMConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package bio.terra.app.configuration;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@EnableConfigurationProperties
+@ConfigurationProperties(prefix = "ecm")
+public class ECMConfiguration {
```
Comment (nmalfroy) [inline src/main/java/bio/terra/app/configuration/ECMConfiguration.java:10]: Curious about your thoughts: should this be `ECM` or `Ecm`?  I always struggle with this.  I'd lean towards the later but I'm struggling to find other example of this

Comment (nmalfroy) [reply src/main/java/bio/terra/app/configuration/ECMConfiguration.java:10]: Eh, you know what, never mind. I see that we have DRS using all caps so let's keep that!  I'll leave my comment in case others go through the same thought

File: src/main/java/bio/terra/app/configuration/ECMConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+  private String instanceUrl;
+
+  public String getInstanceUrl() {
```
Comment (nmalfroy) [inline src/main/java/bio/terra/app/configuration/ECMConfiguration.java]: I think we've ben typically using `basePath` to point to terra service URLs (e.g. Sam and Terra configs). I'd use the same here

File: src/main/java/bio/terra/service/auth/ras/ECMService.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    this.jacksonConverter = jacksonConverter;
+  }
+
+  public PassportApi getPassportApi() {
+    // ECM returns lots of responses as JSON string.
+    // By default, these get intercepted by Spring's raw string converter, but they're quoted, so
+    // they should be parsed as JSON.
+    // N.B. This means that any non-JSON responses won't be parsed correctly, so if ECM ever returns
+    // any other content-type, this would have to be revised. However, that seems unlikely.
+    var restTemplate = new RestTemplate(List.of(jacksonConverter));
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/auth/ras/ECMService.java]: Can you not use the rest template bean that you created above?

File: src/main/java/bio/terra/service/auth/ras/ECMService.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestTemplate;
+
+@Service
+public class ECMService {
+  private final Logger logger = LoggerFactory.getLogger(ECMService.class);
+  private final ECMConfiguration ecmConfiguration;
+  private final MappingJackson2HttpMessageConverter jacksonConverter;
+
+  public ECMService(
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/auth/ras/ECMService.java]: Do we need an `@Autowired`?

File: src/main/java/bio/terra/service/auth/ras/ECMService.java (hunk 2 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/auth/ras/ECMService.java]: I think this is reasonable.  It looks pretty much like how it does for the Sam client

File: src/main/java/bio/terra/service/filedata/DrsService.java [hunk 9]
```diff
@@ -131,6 +139,32 @@ public void close() {
     }
   }
 
+  /*
+  Get/Post DRS Object
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/filedata/DrsService.java]: The extra comment looks a little strange here.  Do we even need it?  Or maybe just a space to separate the comments

File: src/main/java/bio/terra/service/filedata/DrsService.java [hunk 10]
```diff
@@ -131,6 +139,32 @@ public void close() {
...
+   *     only those objects directly contained in the bundle
+   * @return the DRS object for this ID
+   * @throws IllegalArgumentException if there is an issue with the object id
+   * @throws SnapshotNotFoundException if the snapshot for the DRS object cannot be found
+   * @throws TooManyRequestsException if there are too many concurrent DRS lookup requests
+   */
+  public DRSObject lookupObjectByDrsIdPassport(
+      String drsObjectId, DRSPassportRequestModel drsPassportRequestModel) {
+    try (DrsRequestResource r = new DrsRequestResource()) {
+      Snapshot snapshot = lookupSnapshotForDRSObject(drsObjectId);
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/filedata/DrsService.java]: Given the optimization that Trevyn just had to do: we should use the cached object I think

File: src/main/java/bio/terra/service/filedata/DrsService.java [hunk 11]
```diff
@@ -146,69 +180,122 @@ public void close() {
...
-            drsObjectId, // not a flight, so no job id
-            this.getClass().getName(),
-            "snapshotService.retrieveAvailable");
-      } catch (IllegalArgumentException ex) {
-        throw new InvalidDrsIdException("Invalid object id format '" + drsObjectId + "'", ex);
-      } catch (SnapshotNotFoundException ex) {
-        throw new DrsObjectNotFoundException(
-            "No snapshot found for DRS object id '" + drsObjectId + "'", ex);
-      }
+      Snapshot snapshot = lookupSnapshotForDRSObject(drsObjectId);
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/filedata/DrsService.java]: Same comment about the cache

File: src/main/java/bio/terra/service/filedata/DrsService.java [hunk 12]
```diff
@@ -146,69 +180,122 @@ public void close() {
...
+    }
+  }
 
-      if (fsObject instanceof FSFile) {
-        return drsObjectFromFSFile((FSFile) fsObject, drsId.getSnapshotId(), authUser);
-      } else if (fsObject instanceof FSDir) {
-        return drsObjectFromFSDir((FSDir) fsObject, drsId.getSnapshotId());
-      }
+  @VisibleForTesting
+  Snapshot lookupSnapshotForDRSObject(String drsObjectId) {
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/filedata/DrsService.java]: Given how big a snapshot is, I think we should probably avoid using the entire snapshot object if we can avoid it.  I think the `SnapshotCacheResult` object in #1253 might be worthwhile here

File: src/main/java/bio/terra/service/filedata/DrsService.java [hunk 13]
```diff
@@ -146,69 +180,122 @@ public void close() {
...
+  void verifyPassportAuth(UUID snapshotId, DRSPassportRequestModel drsPassportRequestModel) {
+    SnapshotSummaryModel snapshotSummary = snapshotService.retrieveSnapshotSummary(snapshotId);
+    String phsId = snapshotSummary.getPhsId();
+    String consentCode = snapshotSummary.getConsentCode();
+    if (phsId == null || consentCode == null) {
+      throw new InvalidAuthorizationMethod("Snapshot cannot use Ras Passport authorization");
+    }
+    // Pass the passport + phs id + consent code to ECM
+    var criteria = new RASv1Dot1VisaCriterion().consentCode(consentCode).phsId(phsId);
+    criteria.issuer("https://stsstg.nih.gov").type("RASv1Dot1VisaCriterion");
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/filedata/DrsService.java]: These should be in constants

File: src/main/java/bio/terra/service/filedata/DrsService.java (hunk 3 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/filedata/DrsService.java]: I think we need may need to be more explicit.  If you look here https://docs.google.com/document/d/1isNIYTkoOM0TdQ3skcSu79VMh8CCPHq_Zh356VXSVDE/edit#heading=h.ab5n175zvcrt there are the fields that we need.  I'm wondering if a simple formatted message is enough here.

File: src/main/java/bio/terra/service/filedata/DrsService.java [hunk 14]
```diff
@@ -220,41 +307,41 @@ public DRSAccessURL getAccessUrlForObjectId(
...
+      return signAzureUrl(billingProfileModel, fsFile, authUser);
     } else {
       throw new FeatureNotImplementedException("Cloud platform not implemented");
     }
   }
 
-  private Optional<DRSAccessMethod> getAccessMethodMatchingAccessId(
-      String accessId, DRSObject object, TypeEnum methodType) {
+  /*
+  Helper methods
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/filedata/DrsService.java]: Kind of a strange format for the log

**Review by nmalfroy — CHANGES_REQUESTED** [review]
This looks great.  Thank you for breaking up the commits like this, it made it a ton easier to review.  I did have a couple of comments to look into around perf (especially in light of some of the things we ran into yesterday)

File: src/main/resources/api/data-repository-openapi.yaml [hunk 15]
```diff
@@ -2862,6 +2862,75 @@ paths:
...
+      description:  Returns object metadata, and a list of access methods that can be used to fetch object bytes.  Method is a POST to accomodate a JWT GA4GH Passport sent in the formData in order to authorize access.
+      operationId: PostObject
+      parameters:
+        - name: object_id
+          in: path
+          required: true
+          schema:
+            type: string
+      security:
+        - PassportAuth: [ ]
```
Comment (samanehsan) [inline src/main/resources/api/data-repository-openapi.yaml]: What is this `security` field used for? I haven't seen that before!

File: src/main/java/bio/terra/service/filedata/DrsService.java [hunk 16]
```diff
@@ -297,34 +383,39 @@ private DRSAccessURL signGoogleUrl(Snapshot snapshot, DRSAccessMethod accessMeth
...
+      FSFile fsFile, Snapshot snapshot, AuthenticatedUserRequest authUser, boolean passportAuth) {
+    DRSObject fileObject = makeCommonDrsObject(fsFile, snapshot.getId().toString());
 
     List<DRSAccessMethod> accessMethods;
     CloudPlatformWrapper platform = CloudPlatformWrapper.of(fsFile.getCloudPlatform());
     if (platform.isGcp()) {
-      accessMethods = getDrsAccessMethodsOnGcp(snapshotId, fsFile, authUser);
+      String gcpRegion = retrieveGCPSnapshotRegion(snapshot, fsFile);
+      if (passportAuth) {
+        accessMethods = getDrsSignedURLAccessMethods("gcp-passport-", gcpRegion);
```
Comment (samanehsan) [inline src/main/java/bio/terra/service/filedata/DrsService.java]: What do you think about making these prefixes constants?

File: src/test/java/bio/terra/service/filedata/DrsServiceTest.java [hunk 17]
```diff
@@ -188,6 +213,145 @@ public void testLookupNegative() {
         () -> drsService.lookupObjectByDrsId(authUser, azureDrsObjectId, false));
   }
 
+  @Test
+  public void testLookupSnapshot() {
+    drsService.lookupSnapshotForDRSObject(googleDrsObjectId);
```
Comment (samanehsan) [inline src/test/java/bio/terra/service/filedata/DrsServiceTest.java]: Can we assert that this retrieves the expected snapshot?

**Review by samanehsan — APPROVED** [review]
Once the comments from Muscles are addressed this LGTM! Thanks for structuring the commits so clearly! :octocat:

File: src/main/java/bio/terra/service/filedata/DrsService.java (hunk 12 above)
Comment (snf2ye) [reply src/main/java/bio/terra/service/filedata/DrsService.java]: Definitely! Once that PR goes in, I'll add the needed fields to the cache result.

File: src/main/java/bio/terra/service/filedata/DrsService.java (hunk 13 above)
Comment (snf2ye) [reply src/main/java/bio/terra/service/filedata/DrsService.java]: Good point - Are constants in the file good enough or should I pull them to app properties?

File: src/main/resources/api/data-repository-openapi.yaml (hunk 15 above)
Comment (snf2ye) [reply src/main/resources/api/data-repository-openapi.yaml]: Thanks for asking about this! I had copied it from the DRS Spec, but overlooked revisiting what it meant. 
- In general, here's more details on the security field in swagger - https://swagger.io/docs/specification/authentication/
- I'm adding a change to add a security schema to define "passportauth" --> This definition is also copied from the DRS spec. https://github.com/ga4gh/data-repository-service-schemas/blob/master/openapi/data_repository_service.openapi.yaml#L130

Basically, by defining the security for these endpoints, we're overriding [the oauth security descriptor for the rest of the endpoints ](https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/resources/api/data-repository-openapi.yaml#L33)

File: src/main/java/bio/terra/service/filedata/DrsService.java (hunk 13 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/filedata/DrsService.java]: I think in the class is totally reasonable

File: src/main/java/bio/terra/service/filedata/DrsService.java (hunk 9 above)
Comment (snf2ye) [reply src/main/java/bio/terra/service/filedata/DrsService.java]: Removed

File: src/main/java/bio/terra/service/filedata/DrsService.java (hunk 14 above)
Comment (snf2ye) [reply src/main/java/bio/terra/service/filedata/DrsService.java]: Removed

File: src/main/java/bio/terra/service/auth/ras/ECMService.java (hunk 7 above)
Comment (snf2ye) [reply src/main/java/bio/terra/service/auth/ras/ECMService.java]: Oh you're totally right - that cleans this up quite a bit.

File: src/main/java/bio/terra/service/filedata/DrsService.java (hunk 12 above)
Comment (snf2ye) [reply src/main/java/bio/terra/service/filedata/DrsService.java]: Should be added now!

File: src/main/java/bio/terra/service/filedata/DrsService.java (hunk 16 above)
Comment (snf2ye) [reply src/main/java/bio/terra/service/filedata/DrsService.java]: Good point! Should be constants now.

File: src/test/java/bio/terra/service/filedata/DrsServiceTest.java (hunk 17 above)
Comment (snf2ye) [reply src/test/java/bio/terra/service/filedata/DrsServiceTest.java]: Good idea - Added!

File: src/main/resources/api/data-repository-openapi.yaml (hunk 15 above)
Comment (snf2ye) [reply src/main/resources/api/data-repository-openapi.yaml]: As I mentioned in standup, I decided to remove this security option. As planned, the user will just provide the passport as part of the request and we will make this endpoint unauthenticated.

File: src/main/java/bio/terra/app/configuration/ApplicationConfiguration.java (hunk 1 above)
Comment (snf2ye) [reply src/main/java/bio/terra/app/configuration/ApplicationConfiguration.java:406]: Ok! I think actually referencing the RestTemplate bean as you suggested cleaned this up and I'm no longer concerned about it! Thank you!

File: src/main/java/bio/terra/service/filedata/DrsService.java (hunk 3 above)
Comment (snf2ye) [reply src/main/java/bio/terra/service/filedata/DrsService.java]: Thanks for discussing offline. 
Example log:
`[Validate Passport Audit]: Data Repository accessed: TDR, Study/Data set accessed: <redacted>, Date/Time of access: 04-11-2022 16:02:40 EST, ECM Audit Info: {passport_jti=<redacted>, external_user_id=<redacted>, internal_user_id=<redacted>}`

**Review by samanehsan — APPROVED** [review]
The new commits look good to me! :octocat:

File: src/main/java/bio/terra/service/filedata/DrsService.java [hunk 18]
```diff
@@ -187,63 +193,73 @@ public DRSObject lookupObjectByDrsIdPassport(
...
     var request =
         new ValidatePassportRequest()
             .passports(drsPassportRequestModel.getPassports())
             .criteria(List.of(criteria));
     var result = ecmService.validatePassport(request);
 
-    // log audit info
-    logger.info(result.getAuditInfo().toString());
+    var df = new SimpleDateFormat("MM-dd-yyyy HH:mm:ss z");
+    df.setTimeZone(TimeZone.getTimeZone("EST"));
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/filedata/DrsService.java]: Do we want this to be EST?  I think we'd be better served making this UTC

File: src/main/java/bio/terra/service/filedata/DrsService.java [hunk 19]
```diff
@@ -75,6 +77,11 @@ public class DrsService {
 
   private final Logger logger = LoggerFactory.getLogger(DrsService.class);
 
+  private static final String ACCESS_ID_PREFIX_GCP = "gcp-";
+  private static final String ACCESS_ID_PREFIX_AZURE = "azure-";
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/filedata/DrsService.java]: I think this was `az-` originally.  Is it ok that it would change?  Should we leave it as `az-`?

File: src/main/java/bio/terra/service/filedata/DrsService.java [hunk 20]
```diff
@@ -78,7 +78,7 @@ public class DrsService {
   private final Logger logger = LoggerFactory.getLogger(DrsService.class);
 
   private static final String ACCESS_ID_PREFIX_GCP = "gcp-";
-  private static final String ACCESS_ID_PREFIX_AZURE = "azure-";
+  private static final String ACCESS_ID_PREFIX_AZURE = "az-";
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/filedata/DrsService.java:81]: lol...that's what I get for going commit-by-commit :)

**Review by nmalfroy — APPROVED** [review]
Looks great, thanks for making the changes!  Just one thing really about the timezone
